example_scripts/login_system-cache.pwn

```diff
@@ -46,224 +39,286 @@ new Player[MAX_PLAYERS][E_PLAYERS];
+    }
+    return 1;
 }
 
-public OnPlayerSpawn(playerid)
+forward OnLoginTimeout(playerid);
+public OnLoginTimeout(playerid)
 {
-	ResetPlayerMoney(playerid);
-	GivePlayerMoney(playerid, Player[playerid][Money]);
-	return 1;
+    ShowPlayerDialog(playerid, DIALOG_UNUSED, DIALOG_STYLE_MSGBOX, "Login", "You have been kicked for taking too long to login successfully to your account.", "Okay", "");
+    DelayedKick(playerid);
+    return 1;
 }
 
-public OnPlayerDisconnect(playerid,reason)
+forward OnPlayerRegister(playerid);
+public OnPlayerRegister(playerid)
 {
-    g_MysqlRaceCheck[playerid]++;
-	UpdatePlayerData(playerid);
-	return 1;
+    // retrieves the ID generated for an AUTO_INCREMENT column by the sent query
+    Player[playerid][ID] = cache_insert_id();
+
+    ShowPlayerDialog(playerid, DIALOG_UNUSED, DIALOG_STYLE_MSGBOX, "Registration", "Account successfully registered, you have been automatically logged in.", "Okay", "");
+
+    Player[playerid][IsLoggedIn] = true;
+
+    SetSpawnInfo(playerid, NO_TEAM, 0, 1958.3783, 1343.1572, 15.3746, 270.1425, 0, 0, 0, 0, 0, 0);
+    SpawnPlayer(playerid);
+    return 1;
 }
 
+forward _KickPlayerDelayed(playerid);
+public _KickPlayerDelayed(playerid)
+{
+    Kick(playerid);
+    return 1;
+}
 
-/*
- * functions
- */
+
+//-----------------------------------------------------
 
 AssignPlayerData(playerid)
 {
-	Player[playerid][ID] = cache_get_field_content_int(0, "id");
-	cache_get_field_content(0, "password", Player[playerid][Password], g_SQL, 129);
-	Player[playerid][Money] = cache_get_field_content_int(0, "money");
-	return 1;
+    cache_get_value_int(0, "id", Player[playerid][ID]);
+    cache_get_value_int(0, "kills", Player[playerid][Kills]);
+    cache_get_value_int(0, "deaths", Player[playerid][Deaths]);
+    return 1;
 }
 
-UpdatePlayerData(playerid)
+DelayedKick(playerid, time = 500)
 {
-	if(Player[playerid][IsLoggedIn] == false)
-	    return 0;
-	    
-	new query[128];
-	mysql_format(g_SQL, query, sizeof(query), "UPDATE `players` SET `money` = '%d' WHERE `id` = '%d' LIMIT 1", Player[playerid][Money], Player[playerid][ID]);
-	mysql_tquery(g_SQL, query);
-	return 1;
+    SetTimerEx("_KickPlayerDelayed", time, false, "d", playerid);
+    return 1;
 }
 
 SetupPlayerTable()
 {
-	mysql_query(g_SQL, "CREATE TABLE IF NOT EXISTS `players` (`id` int(11) NOT NULL auto_increment PRIMARY KEY,`username` varchar(30) NOT NULL,`password` varchar(130) NOT NULL,`money` int(10) NOT NULL default '0')", false);
-	return 1;
+    mysql_tquery(g_SQL, "CREATE TABLE IF NOT EXISTS `players` (`id` int(11) NOT NULL AUTO_INCREMENT, `username` varchar(24) NOT NULL, `password` char(64) NOT NULL, `salt` char(16) NOT NULL, `kills` int(11) NOT NULL DEFAULT '0', `deaths` int(11) NOT NULL DEFAULT '0', PRIMARY KEY (`id`), UNIQUE KEY `username` (`username`))");
+    return 1;
 }
 
-
-DelayedKick(playerid, time=500)
+UpdatePlayerData(playerid)
```

Comment:
Would be great if you split that function up into `UpdatePlayerKills()` and `UpdatePlayerDeaths()`, or maybe even `SetPlayerKills(kills)` and `SetPlayerDeaths(deaths)` (and `Get` functions). If people take this script as their gamemode base, the `UPDATE` query will get very long by time. A value in a MySQL table should directly be updated if the corresponding PAWN variable changes.